### PR TITLE
Sortable list

### PR DESCRIPTION
**Pull request by kristofferj**
DefaultList has now a sortable prop. 

Some issues are still here.
The DefaultList with sortable-prop to true, seems to re-render on onSelect. 
When onSortStart is provided to DefaultList, chrome browser crashes. 
Check the storybook for examples and testing.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,8 +23,10 @@
     "sanity-plugin"
   ],
   "dependencies": {
+    "babel-eslint": "^7.1.0",
```
Comment (bjoerge) [inline package.json]: Think this should go in `devDependencies`

File: package.json [hunk 2]
```diff
@@ -23,8 +23,10 @@
...
     "chance": "^1.0.4",
     "dialog-polyfill": "^0.4.3",
+    "eslint-plugin-react": "^6.6.0",
```
Comment (bjoerge) [inline package.json]: Same

File: package.json [hunk 3]
```diff
@@ -38,7 +40,7 @@
     "react-ink": "^5.1.1",
     "react-sortable-hoc": "^0.1.0",
     "stylelint": "^7.1.0",
-    "stylelint-config-standard": "^12.0.0"
+    "stylelint-config-standard": "^14.0.0"
```
Comment (bjoerge) [inline package.json]: And this

File: src/lists/DefaultList.js [hunk 4]
```diff
@@ -66,36 +69,78 @@ export default class DefaultList extends React.Component {
...
 
-    const {items, className, selectedItem, highlightedItem, renderItem, scrollable, decoration} = this.props
+    const {items, className, scrollable, sortable, useDragHandle, onSortStart, onSortEnd, onSortMove} = this.props
+
+    const SortableItem = sortable && SortableElement(({value, index}) => {
+      return this.renderListitem(value, this.props, index)
+    })
 
-    const ListItemContainer = this.props.ListItemContainer || ListItem
+    const SortableList = sortable && SortableContainer(({sortableItems}) => {
```
Comment (bjoerge) [inline src/lists/DefaultList.js]: Could this be moved to outside of the class? I'm worried about the performance impact of creating a component class on every render here.

File: src/lists/DefaultList.js [hunk 5]
```diff
@@ -66,36 +69,78 @@ export default class DefaultList extends React.Component {
...
+            !sortable && <ul className={styles.list} ref={this.setListContainer}>
+              {
+                items && items.map((item, index) => {
+                  return this.renderListitem(item, this.props, index)
+                })
+              }
+            </ul>
+          }
+          {
+            sortable && SortableList && <SortableList
```
Comment (bjoerge) [inline src/lists/DefaultList.js]: IMO, wrapping the `<SortableList` in parens would be more readable here:
```
sortable && SortableList && (
  <SortableList
  ...
  />
)}
```

File: src/lists/items/DefaultItem.js [hunk 6]
```diff
@@ -12,7 +12,7 @@ export default class DefaultListItem extends React.Component {
     selected: PropTypes.bool,
     highlighted: PropTypes.bool,
     scrollIntoView: PropTypes.func,
-    decoration: PropTypes.oneOf(['default', 'zebra-stripes'])
+    decoration: PropTypes.oneOf(['default', 'zebra-stripes', 'divider'])
```
Comment (bjoerge) [inline src/lists/items/DefaultItem.js]: 👍

File: src/lists/story.js [hunk 7]
```diff
@@ -28,29 +27,54 @@ const defaultItems = range(100).map((item, i) => {
...
-  handleOnSort = () => {
-    // console.log('on sort')
-  }
-
   handleOnSortEnd = ({oldIndex, newIndex}) => {
     const {items} = this.state
     this.setState({
       items: arrayMove(items, oldIndex, newIndex)
     })
+    this.props.onSortEnd()
```
Comment (bjoerge) [inline src/lists/story.js]: How will the consumer get a hold of the new order/array value here? Shouldn't this emit the items?

Comment (bjoerge) [inline src/lists/story.js]: Also, I in the form builder we somehow need to translate this to a patch, e.g. "move item x to position y"

Comment (bjoerge) [inline src/lists/story.js]: Oh, I just realized this is in story-file. All good then. Maybe rename the class to SortableComponentStory or something similar to be more clear?

**Comment by bjoerge** [comment]
Looks good! There's a few issues that we need to take a closer look at before merging. There's also a merge conflict that needs to be resolved.

File: src/lists/DefaultList.js (hunk 4 above)
Comment (kristofferj) [inline src/lists/DefaultList.js]: Hm. sortableItems and ListItemContainer is a prop on DefaultList. Are they available outside the class?

Comment (bjoerge) [inline src/lists/DefaultList.js]: No, but they can be passed on as props to the created SortableContainer like you do with other props (i believe)
Something like this:

```jsx
const SortableItemContainer = SortableContainer(function Item(props) {
  const {SortableItem, sortableItems} = props
  return (
    <ul className={styles.sortableList}>
      {sortableItems.map((value, index) => {
          return (
            <SortableItem key={`sortableItem-${index}`} index={index} value={value}/>
          )
        })
      }
    </ul>
  )
})
```

Usage:
```jsx
<SortableContainer sortableItems={...} ListItemContainer={this.props.ListItemContainer || ListItem} .../>
```

**Comment by kristofferj** [comment]
@bjoerge About time to merge?

**Pull request merged**